File: app/Quotes/bs/src/widgets/wserver_ids.ml
(*  Copyright 30-Jul-2018 ºDeme
    GNU General Public License - V3 <http://www.gnu.org/licenses/>
*)

open Ui
open I18n
open Wserver_id

let mk nick servers =
  let fields = List.map (Wserver_id.mk nick) servers in
  let td field =
    q "td" [Klass "frame"; Style "text-align:center;width:33%;"][
      Wrule.mk_small (server field);
      widget field]
  in
  q "div" [Style "text-align:center;"] [
    Wrule.mk_big (i18 "Server codes");
    link (fun _ -> List.iter (fun field -> download field) fields) |>
      Domo.set [Klass "link"; Text "Test all"];
    q "table" [Style "width:100%"; Att ("cellspacing", "10")][
      q "tr" [] (List.map td fields)]]

File: app/Quotes201812/js/src/wgs/WserverIds.js
// Copyright 13-Sept-2018 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// eslint-disable-next-line
import Main from "../Main.js";
import Ui from "../dmjs/Ui.js";
// eslint-disable-next-line
import Tp from "../dmjs/Tp.js";
import {_} from "../I18n.js";
// eslint-disable-next-line
import Domo from "../dmjs/Domo.js";
import Wrule from "./Wrule.js";
import WserverId from "./WserverId.js";

const $ = Ui.$;

/** Widget for reading a server identifier */
export default class WserverIds {
  /**
   * @param {!Main} main
   * @param {!Array<Tp<string, string>>} serverIdCodes They are pairs
   *         <serverId, nickCode>
   * @param {string} nickId
   */
  constructor (main, serverIdCodes, nickId) {
    this._main = main;

    this._wservers = serverIdCodes
      .map(idCode => new WserverId(main, idCode.e1, nickId, idCode.e2));
  }

  test () {
    for (const ws of this._wservers) {
      ws.test();
    }
  }

  /** @return {!Domo} */
  wg () {
    const self = this;
    return $("div").style("text-align:center;")
      .add(Wrule.mkBig(_("Server codes")))
      .add(Ui.link(() => {
        self.test();
      }).klass("link").text(_("Test all")))
      .add($("table").style("width:100%").att("cellspacing", 10)
        .add($("tr")
          .adds(self._wservers.map(ws =>
            $("td").klass("frame").style("text-align:center;width:33%;")
              .add(Wrule.mkSmall(ws.serverId))
              .add(ws.wg())
          ))))
    ;

  }
}
